include/misc/math_extras_impl.h

```diff
@@ -0,0 +1,187 @@
+#define use_builtin(x) HAS_BUILTIN(x)
+#endif
+
+#if use_builtin(__builtin_add_overflow)
+static inline bool u32_add_overflow(u32_t a, u32_t b, u32_t *result)
+{
+	return __builtin_add_overflow(a, b, result);
+}
+
+static inline bool u64_add_overflow(u64_t a, u64_t b, u64_t *result)
+{
+	return __builtin_add_overflow(a, b, result);
+}
+
+static inline bool size_add_overflow(size_t a, size_t b, size_t *result)
+{
+	return __builtin_add_overflow(a, b, result);
+}
+#else /* !use_builtin(__builtin_add_overflow) */
+static inline bool u32_add_overflow(u32_t a, u32_t b, u32_t *result)
+{
+	u32_t c = a + b;
+
+	*result = c;
+
+	return c < a;
+}
+
+static inline bool u64_add_overflow(u64_t a, u64_t b, u64_t *result)
+{
+	u64_t c = a + b;
+
+	*result = c;
+
+	return c < a;
+}
+
+static inline bool size_add_overflow(size_t a, size_t b, size_t *result)
+{
+	size_t c = a + b;
+
+	*result = c;
+
+	return c < a;
+}
+#endif /* use_builtin(__builtin_add_overflow) */
+
+#if use_builtin(__builtin_mul_overflow)
+static inline bool u32_mul_overflow(u32_t a, u32_t b, u32_t *result)
+{
+	return __builtin_mul_overflow(a, b, result);
+}
+
+static inline bool u64_mul_overflow(u64_t a, u64_t b, u64_t *result)
+{
+	return __builtin_mul_overflow(a, b, result);
+}
+
+static inline bool size_mul_overflow(size_t a, size_t b, size_t *result)
+{
+	return __builtin_mul_overflow(a, b, result);
+}
+#else /* !use_builtin(__builtin_mul_overflow) */
+static inline bool u32_mul_overflow(u32_t a, u32_t b, u32_t *result)
+{
+	u32_t c = a * b;
+
+	*result = c;
+
+	return a > 0 && c / a != b;
+}
+
+static inline bool u64_mul_overflow(u64_t a, u64_t b, u64_t *result)
+{
+	u64_t c = a * b;
+
+	*result = c;
+
+	return a > 0 && c / a != b;
+}
+
+static inline bool size_mul_overflow(size_t a, size_t b, size_t *result)
+{
+	size_t c = a * b;
+
+	*result = c;
+
+	return a > 0 && c / a != b;
+}
+#endif /* use_builtin(__builtin_mul_overflow) */
+
+#if use_builtin(__builtin_clz)
+static inline int u32_count_leading_zeros(u32_t x)
+{
+	return x == 0 ? 32 : __builtin_clz(x);
+}
+#else /* !use_builtin(__builtin_clz) */
+static inline int u32_count_leading_zeros(u32_t x)
+{
+	int b = 0;
+
+	while (b < 32 && (x >> 31) == 0) {
+		b++, x <<= 1;
+	}
+
+	return b;
+}
+#endif /* use_builtin(__builtin_clz) */
+
+#if use_builtin(__builtin_clzll)
+static inline int u64_count_leading_zeros(u64_t x)
+{
+	return x == 0 ? 64 : __builtin_clzll(x);
+}
+#else /* !use_builtin(__builtin_clzll) */
+static inline int u64_count_leading_zeros(u64_t x)
+{
+	if (x == (u32_t)x) {
+		return 32 + u32_count_leading_zeros((u32_t)x);
+	} else {
+		return u32_count_leading_zeros(x >> 32);
+	}
+}
+#endif /* use_builtin(__builtin_clzll) */
+
+#if use_builtin(__builtin_ctz)
+static inline int u32_count_trailing_zeros(u32_t x)
+{
+	return x == 0 ? 32 : __builtin_ctz(x);
+}
+#else /* !use_builtin(__builtin_ctz) */
+static inline int u32_count_trailing_zeros(u32_t x)
+{
+	int b = 0;
+
+	while (b < 32 && (x & 1) == 0) {
+		b++, x >>= 1;
```

Comment:
So... frankly I don't personally care, but that use of the comma operator is needless and I'm all but certain someone's coding style document (@ceolin can tell you all about the world he's been stuck in) disallows this.   Actually given the structure, maybe this should just be a `for()`  (where, heh, you could use the comma in a well-justified way because it has to fit in the final expression).